stan/math/rev/core/vari.hpp

```diff
@@ -4,12 +4,53 @@
 #include <stan/math/rev/core/chainable_alloc.hpp>
 #include <stan/math/rev/core/chainablestack.hpp>
 #include <ostream>
+#include <type_traits>
 
 namespace stan {
 namespace math {
 
 // forward declaration of var
-class var;
+template <typename T>
+class var_value;
+
+/**
+ * Pure virtual class that all `vari_value` and it's derived classes inherit.
+ */
+class vari_base {
+ public:
+  /**
+   * Apply the chain rule to this variable based on the variables
+   * on which it depends.
+   */
+  virtual void chain() {}
+
+  /**
+   * Set the adjoint value of this variable to 0.  This is used to
+   * reset adjoints before propagating derivatives again (for
+   * example in a Jacobian calculation).
+   */
+  virtual void set_zero_adjoint() = 0;
+
+  /**
+   * Throw an illegal argument exception.
+   *
+   * <i>Warning</i>: Destructors should never called for var objects.
+   *
+   * @throw Logic exception always.
+   */
+  virtual ~vari_base() {}
+};
+
+namespace internal {
+/**
+ * Promote integral types to double
+ * @tparam Val Any type
+ */
+template <typename Val>
+using floating_point_promoter
+    = std::conditional_t<std::is_integral<std::decay_t<Val>>::value, double,
+                         std::decay_t<Val>>;
```

Comment:
[optional] I think a better name for this would be `promote_floating_point`. Also it is used in multiple files, so I see no reason for it being in internal namespace. It could be instead in its own file in meta folder.

Reply:
Reasonable!

Reply:
Actually I'm going to get rid of this because `boost::promote_args` does the same thing as what I'm doing here